File: src/lib/actor/gen_server.ts
import { fx, type Msg, type Pid, type ProcFn, type Ref } from "./types.ts";

export type Callback<S = unknown, Req = unknown, Rep = unknown> = {
  init: () => S;
  handleCall: (req: Req, from: Pid, state: S) => { reply: Rep; state: S };
  handleCast?: (req: Req, state: S) => S;
};

const L = "gen_server.ml:loop";

export function startLink<S, Req, Rep>(
  name: string,
  cb: Callback<S, Req, Rep>,
): ProcFn {
  return function* () {
    yield fx.register(name, "gen_server.ml:register");
    let state = cb.init();
    while (true) {
      const msg = (yield fx.receive(
        (m) => m.t === "Call" || m.t === "Cast" || m.t === "Crash",
        L,
      )) as Msg;
      if (msg.t === "Crash") throw new Error("gen_server abort");
      if (msg.t === "Call") {
        const out = cb.handleCall(msg.req as Req, msg.from, state);
        state = out.state;
        // OTP 24+ gen:reply sends to the alias, not the Pid.
        yield fx.send(
          { alias: msg.ref },
          { t: "Reply", ref: msg.ref, reply: out.reply },
          "gen_server.ml:reply",
        );
      } else if (msg.t === "Cast") {
        state = cb.handleCast
          ? cb.handleCast(msg.req as Req, state)
          : state;
      }
    }
  };
}

/**
 * OTP 24+ gen:call — monitor with `{alias, demonitor}`, send, receive
 * Reply | DOWN. Reply is addressed to the alias. Timeout demonitors
 * (alias dies); a Reply already in the mailbox is still taken; a
 * later send to the alias drops.
 */
export function* call(
  to: Pid,
  req: unknown,
  loc = "gen_server.ml:call",
  timeout = 4000,
): Generator<import("./types.ts").Effect, unknown, unknown> {
  const me = (yield fx.self(loc)) as Pid;
  const ref = (yield fx.monitor(to, loc, "demonitor")) as Ref;
  yield fx.send(to, { t: "Call", from: me, ref, req }, loc);
  const m = (yield fx.receive(
    (x) =>
      (x.t === "Reply" && x.ref === ref) || (x.t === "DOWN" && x.ref === ref),
    loc,
    timeout,
    ref,
  )) as Msg;
  if (m.t === "Timeout") {
    yield fx.demonitor(ref, loc, true);
    const leftover = (yield fx.receive(
      (x) => x.t === "Reply" && x.ref === ref,
      loc,
      0,
    )) as Msg;
    if (leftover.t === "Reply") return leftover.reply;
    throw new Error("timeout");
  }
  yield fx.demonitor(ref, loc, true);
  if (m.t === "Reply") return m.reply;
  if (m.t === "DOWN") throw new Error(m.reason);
  throw new Error("timeout");
}

export function echo(): Callback<null, unknown, unknown> {
  return {
    init: () => null,
    handleCall: (req, _from, state) => ({
      reply: req === "ping" ? "pong" : req,
      state,
    }),
  };
}

File: ocaml/gen_server.ml
(* gen_server.ml — OTP gen_server as a functor, not a -behaviour attribute.
   OTP 24+ call is monitor ~alias:true + send + selective receive of
   Reply | DOWN.  gen:reply sends to the alias, not the Pid.  Timeout
   deactivates the alias; a late Reply is dropped the same way a send
   to a dead Pid is dropped. *)

open Actor

type msg +=
  | Call of pid * ref * msg
  | Reply of ref * msg
  | Cast of msg
  | Ping
  | Pong
  | Nudge

module type Callback = sig
  type state
  val init : unit -> state
  val handle_call : msg -> pid -> state -> msg * state
  val handle_cast : msg -> state -> state
end

module Make (C : Callback) = struct
  let start_link name =
    spawn ~name ~link:true @@ fun () ->
      register name;
      let rec loop state =
        match receive (function Call _ | Cast _ | Crash -> true | _ -> false) () with
        | Crash -> failwith "gen_server abort"
        | Call (from, r, req) ->
            let reply, state = C.handle_call req from state in
            send_alias r (Reply (r, reply));
            loop state
        | Cast req -> loop (C.handle_cast req state)
        | _ -> loop state
      in
      loop (C.init ())

  (* OTP 24+ gen:call — monitor with {alias, demonitor}. *)
  let call pid req =
    let me = self () in
    let r = monitor ~alias:true pid in
    send pid (Call (me, r, req));
    match
      receive ~timeout:4.0
        (function
          | Reply (x, _) when x = r -> true
          | Down (x, _, _) when x = r -> true
          | _ -> false)
        ()
    with
    | Reply (_, rep) ->
        demonitor r ~flush:true;
        rep
    | Down (_, _, reason) ->
        demonitor r ~flush:true;
        failwith reason
    | Timeout ->
        demonitor r ~flush:true;
        (match
           receive ~timeout:0.
             (function Reply (x, _) when x = r -> true | _ -> false)
             ()
         with
         | Reply (_, rep) -> rep
         | _ -> failwith "timeout")
    | _ ->
        demonitor r ~flush:true;
        failwith "timeout"

  let cast pid req = send pid (Cast req)
end

module Echo = Make (struct
  type state = unit
  let init () = ()
  let handle_call req _from state =
    match req with
    | Ping -> (Pong, state)
    | other -> (other, state)
  let handle_cast _req state = state
end)
